src/dist/replication/meta_server/meta_server_failure_detector.cpp

```diff
@@ -263,6 +263,10 @@ void meta_server_failure_detector::on_ping(const fd::beacon_msg& beacon, ::dsn::
     {
         ack.is_master = false;
         ack.primary_node = get_primary();
+        {
```

Comment:
Here is what I have in my mind during master switch, which seems inconsistent with what you described above:
- when the switch happens within grace period, then R should not consider itself offline.
- A non-master meta server should not handle the Pings from R, and it considers all Rs alive when it switches itself from non-master to master. One problem is that it may not know all Rs during switch, which breaks the perfect FD invariance. But this is due to the assumption that if a new master does not know R, then R is dead. We can fix it by adding one exception: when a new master is within its first grace period, even a R is not known, it is considered alive from the meta server's point of view. The current fix relies on the old (wrong) pings from the Rs, which is unreliable (as the R may never send a ping to this new meta server before).